Project/Source/Components/ComponentParticleSystem.cpp

```diff
@@ -593,43 +648,41 @@ void ComponentParticleSystem::Draw() {
 				float4x4 newModelMatrix = currentParticle.model.LookAt(rotatePart.Col(2), float3::unitY, rotatePart.Col(1), float3::unitY);
 				modelMatrix = float4x4::FromTRS(currentParticle.position, newModelMatrix.RotatePart(), currentParticle.scale);
 			} else if (billboardType == BillboardType::VERTICAL) {
-				// TODO: Implement it
-				modelMatrix = currentParticle.model;
+				float3 cameraPos = App->camera->GetActiveCamera()->GetFrustum()->Pos();
+				float3 cameraDir = (float3(cameraPos.x, currentParticle.position.y, cameraPos.z) - currentParticle.position).Normalized();
+				float4x4 newModelMatrix = currentParticle.model.LookAt(rotatePart.Col(2), cameraDir, rotatePart.Col(1), float3::unitY);
+				modelMatrix = float4x4::FromTRS(currentParticle.position, newModelMatrix.RotatePart(), currentParticle.scale);
 			}
 
+			float4x4* proj = &App->camera->GetProjectionMatrix();
+			float4x4* view = &App->camera->GetViewMatrix();
+
 			glUniformMatrix4fv(glGetUniformLocation(program, "model"), 1, GL_TRUE, modelMatrix.ptr());
 			glUniformMatrix4fv(glGetUniformLocation(program, "view"), 1, GL_TRUE, view->ptr());
 			glUniformMatrix4fv(glGetUniformLocation(program, "proj"), 1, GL_TRUE, proj->ptr());
 
-			if (!isRandomFrame) {
-				if (App->time->HasGameStarted()) {
-					currentParticle.currentFrame += animationSpeed * App->time->GetDeltaTime();
-				} else {
-					currentParticle.currentFrame += animationSpeed * App->time->GetRealTimeDeltaTime();
-				}
+			float4 color = float4::one;
+			if (colorOverLifetime) {
+				float factor = 1 - currentParticle.life / life; // Life decreases from Life to 0
+				gradient.getColorAt(factor, color.ptr());
 			}
 
-			glActiveTexture(GL_TEXTURE0);
-			glUniform1i(glGetUniformLocation(program, "diffuse"), 0);
+			glUniform1i(glGetUniformLocation(program, "diffuseMap"), 0);
+			glUniform1i(glGetUniformLocation(program, "hasDiffuseMap"), hasDiffuseMap);
+			glUniform4fv(glGetUniformLocation(program, "inputColor"), 1, color.ptr());
```

Comment:
This could be a good moment to migrate the GetUniformLocation calls through the new Program classes. Look at the way this was implemented in ComponentMeshRenderer for an example.